### PR TITLE
Add SRFI-118 - Simple adjustable-size strings

### PR DESCRIPTION
**Pull request by jpellegrini**
This adds two procedures, written in C and placed in
str.c:

* string-append!
* string-replace!

An auxiliary function get_substring_size() was added also.

### CODE REVIEW

**Comment by jpellegrini** [comment]
The first argumet may be a literal, because it seems to work:
```
stklos> (string-append! "a" "b")
"ab"
```
Then, there is SRFI-185, which seems to be trivially implemented on top of this one, and offers macros that require the first argument to be a place...

**Comment by egallesio** [comment]
Great, another interesting SRFI!

> The first argumet may be a literal, because it seems to work in STklos:
> 
> ```
> stklos> (string-append! "a" "b")
> "ab"
> ```
> 

I think that we must verify that the fist string is not a constant one. Otherwise, we could have 

```scheme
stklos>  (define hello "Hello")
;; hello
stklos>   (define s1 hello)
;; s1
stklos>   (define s2 hello)
;; s2
stklos>   (string-append! s1 " world.")
"Hello world."
stklos> (list s1 s2)
("Hello world." "Hello world.")
```
that is the constant string `hello` whose value is changed! 

Testing that `str` is constant can be done with 

```c
if (BOXED_INFO(str) & STRING_CONST)   error_change_const_string(str);
```
What do you think?

File: src/str.c [hunk 1]
```diff
@@ -677,6 +687,212 @@ DEFINE_PRIMITIVE("string-append", string_append, vsubr, (int argc, SCM* argv))
...
+      total_length += STRING_LENGTH(*(argv-i));
+    } else { /* we have already checked, it is a char! */
+      p += STk_char2utf8(CHARACTER_VAL(*(argv-i)),p);
+      /*      p += STk_utf8_char_bytes_needed(CHARACTER_VAL(p));*/
+      total_length ++;
+    }
+  }
+
+  STRING_LENGTH(q) = total_length;
+  fflush(stdout);
```
Comment (egallesio) [inline src/str.c:751]: could be deleted?

File: src/str.c [hunk 2]
```diff
@@ -1381,6 +1597,8 @@ int STk_init_string(void)
   ADD_PRIMITIVE(strgei);
   ADD_PRIMITIVE(substring);
   ADD_PRIMITIVE(string_append);
+  ADD_PRIMITIVE(string_append_adjsize);
```
Comment (egallesio) [inline src/str.c:1600]: Could you use please dstring_append (for destructive string-append)? This is how function are named in C when they modify their first parameter and a non destructive version of the function exists.

Comment (jpellegrini) [reply src/str.c:1600]: Sure! I'll work on this later (maybe tomorrow in the morning)

File: src/str.c (hunk 1 above)
Comment (jpellegrini) [reply src/str.c:751]: Ha ha! Sorry, that's from debugging, I forgot to delete. Will do that.

**Comment by jpellegrini** [comment]
> I think that we must verify that the fist string is not a constant one. Otherwise, we could have

Yes, I misread the SRFI! It says "The string must be a variable-size mutable string".

I'll fix that.

**Comment by jpellegrini** [comment]
Ok @egallesio !

I think it's all fixed now -- the debugging leftovers, the mutable-setting bug, and the function name (it's `string_d...` as I see in the source code).

See if it's ok.

**Comment by egallesio** [comment]
Once again, **great job** @jpellegrini !!!

**Pull request merged**